Sources/Class-Data_Validator.php

```diff
@@ -0,0 +1,1326 @@
+	 *
+	 * @param string $field
+	 * @param mixed[] $input
+	 * @param mixed[]|null $validation_parameters array or null
+	 */
+	protected function _validate_integer($field, $input, $validation_parameters = null)
+	{
+		if (!isset($input[$field]))
+			return;
+
+		$filter = filter_var($input[$field], FILTER_VALIDATE_INT);
+
+		if ($filter === false && version_compare(PHP_VERSION, 5.4, '<') && ($input[$field] === '+0' || $input[$field] === '-0'))
+		{
+			$filter = true;
+		}
+
+		if ($filter === false)
+		{
+			return array(
+				'field' => $field,
+				'input' => $input[$field],
+				'function' => __FUNCTION__,
+				'param' => $validation_parameters
+			);
+		}
+	}
+
+	/**
+	 * boolean ... Determine if the provided value is a boolean
+	 *
+	 * Usage: '[key]' => 'boolean'
+	 *
+	 * @param string $field
+	 * @param mixed[] $input
+	 * @param mixed[]|null $validation_parameters array or null
+	 */
+	protected function _validate_boolean($field, $input, $validation_parameters = null)
+	{
+		if (!isset($input[$field]))
+			return;
+
+		$filter = filter_var($input[$field], FILTER_VALIDATE_BOOLEAN, FILTER_NULL_ON_FAILURE);
+
+		// Fixed in php 7 and later in php 5.6.27 https://bugs.php.net/bug.php?id=67167
+		if (version_compare(PHP_VERSION, '5.6.27', '>='))
+		{
+			$filter = $filter;
+		}
+		if (version_compare(PHP_VERSION, 5.4, '<') && $filter === null && ($input[$field] === false || $input[$field] === ''))
+		{
+			$filter = false;
+		}
+		if ($filter === false && is_object($input[$field]) && method_exists($input[$field], '__tostring') === false)
+		{
+			$filter = null;
+		}
+
+		if ($filter === null)
+		{
+			return array(
+				'field' => $field,
+				'input' => $input[$field],
+				'function' => __FUNCTION__,
+				'param' => $validation_parameters
+			);
+		}
+	}
+
+	/**
+	 * float ... Determine if the provided value is a valid float
+	 *
+	 * Usage: '[key]' => 'float'
+	 *
+	 * @param string $field
+	 * @param mixed[] $input
+	 * @param mixed[]|null $validation_parameters array or null
+	 */
+	protected function _validate_float($field, $input, $validation_parameters = null)
+	{
+		if (!isset($input[$field]))
+			return;
+
+		if (filter_var($input[$field], FILTER_VALIDATE_FLOAT) === false)
+		{
+			return array(
+				'field' => $field,
+				'input' => $input[$field],
+				'function' => __FUNCTION__,
+				'param' => $validation_parameters
+			);
+		}
+	}
+
+	/**
+	 * valid_url ... Determine if the provided value is a valid-ish URL
+	 *
+	 * Usage: '[key]' => 'valid_url'
+	 *
+	 * @param string $field
+	 * @param mixed[] $input
+	 * @param mixed[]|null $validation_parameters array or null
+	 */
+	protected function _validate_valid_url($field, $input, $validation_parameters = null)
+	{
+		if (!isset($input[$field]))
+			return;
+
+		if (!preg_match('`^(https{0,1}?:(//([a-z0-9\-._~%]+)(:[0-9]+)?(/[a-z0-9\-._~%!$&\'()*+,;=:@]+)*/?))(\?[a-z0-9\-._~%!$&\'()*+,;=:@/?]*)?(\#[a-z0-9\-._~%!$&\'()*+,;=:@/?]*)?$`', $input[$field], $matches))
```

Comment:
Why is this using a regex rather than `filter_var($input[$field], FILTER_VALIDATE_URL)`?

Also, it might be valuable to add a `_validate_valid_iri()` function to handle user entered URLs that include non-ASCII characters. A simplified version of [this regex](https://github.com/SimpleMachines/SMF2.1/blob/release-2.1/Sources/Subs.php#L1886) should be helpful for that.